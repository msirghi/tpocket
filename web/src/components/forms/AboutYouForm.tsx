import React, { useState } from 'react';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import { CheckboxWithLabel, TextField } from 'formik-material-ui';
import { Formik, Form, Field } from 'formik';
import { Button, LinearProgress } from '@material-ui/core';
import { isPasswordWeak, validateEmailStr } from '../../utils/extensions';
import { useRegisterMutation } from '../../generated/graphql';
import { AlertMessage } from '../alerts/AlertMessage';
import { AlertType } from '../../commons/enums';
import { AES } from 'crypto-js';

type Props = {
  handleNext: () => void;
  handleRegister: (id: number) => void;
};

interface Values {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  repeatPassword: string;
  checked: boolean;
}

export const AboutYouForm: React.FC<Props> = ({ handleNext, handleRegister }) => {
  const [register] = useRegisterMutation();
  const [error, setError] = useState<string | null>();

  const onSubmit = async (values: Values, setSubmitting: Function) => {
    const { email, password, firstName, lastName } = values;
    setSubmitting(true);
    setError(null);
    try {
      const response = await register({
        variables: {
          email,
          firstName,
          lastName,
          password
        }
      });
      handleNext();
      if (response && response.data) {
        const userId = response.data.register.id;
        handleRegister(+userId);
        const encrypted = AES.encrypt(userId, process.env.REACT_APP_TEMP_ID_KEY!);
        localStorage.setItem('tempId', encrypted.toString());
      }
    } catch (e) {
      setError(e.graphQLErrors.map((x: { message: string }) => x.message)[0]);
    }
  };

  return (
    <Formik
      initialValues={{
        firstName: '',
        lastName: '',
        email: '',
        password: '',
        repeatPassword: '',
        checked: false
      }}
      validateOnChange
      validate={(values) => {
        const errors: Partial<Values> = {};
        const { email, password, repeatPassword, firstName, lastName } = values;
        if (!email) {
          errors.email = 'Required.';
        } else if (validateEmailStr(email)) {
          errors.email = 'Invalid email address.';
        }

        if (isPasswordWeak(password)) {
          errors.password = 'Password is too weak.';
        } else if (password !== repeatPassword) {
          errors.password = 'Passwords do not match.';
        }

        if (firstName.length < 2) {
          errors.firstName = 'Enter a valid first name';
        }
        if (lastName.length < 2) {
          errors.lastName = 'Enter a valid last name';
        }

        return errors;
      }}
      onSubmit={async (values, { setSubmitting }) => await onSubmit(values, setSubmitting)}
    >
      {({ submitForm, isSubmitting, values }) => (
        <Form>
          {isSubmitting && <LinearProgress />}
          <Typography variant='h6' gutterBottom>
            About You
          </Typography>
          <Grid container spacing={3}>
            <Grid item xs={12} sm={6}>
              <Field
                component={TextField}
                required
                name='firstName'
                label='First name'
                fullWidth
                autoComplete='given-name'
              />
            </Grid>
            <Grid item xs={12} sm={6}>
              <Field
                component={TextField}
                id='lastName'
                name='lastName'
                label='Last name'
                fullWidth
                autoComplete='family-name'
              />
            </Grid>
            <Grid item xs={12}>
              <Field
                component={TextField}
                required
                id='email'
                name='email'
                label='Email'
                fullWidth
                autoComplete='email'
              />
            </Grid>
            <Grid item xs={12} sm={12}>
              <Field
                component={TextField}
                required
                type='password'
                id='password'
                name='password'
                label='Password'
                fullWidth
              />
            </Grid>
            <Grid item xs={12} sm={12}>
              <Field
                component={TextField}
                required
                type='password'
                id='repeatPassword'
                name='repeatPassword'
                label='Repeat your password'
                fullWidth
              />
            </Grid>
            <Grid item xs={12}>
              <Field
                component={CheckboxWithLabel}
                type='checkbox'
                Label={{ label: 'Accept Terms and Conditions' }}
                name='checked'
              />
            </Grid>
            {error && (
              <Grid item xs={12}>
                <AlertMessage type={AlertType.ERROR} message={error} />
              </Grid>
            )}
            <Grid item xs={12} className='row row-center'>
              <Button
                variant='contained'
                color='primary'
                fullWidth
                disabled={
                  !values.email ||
                  !values.firstName ||
                  !values.lastName ||
                  !values.password ||
                  !values.repeatPassword ||
                  !values.checked ||
                  isSubmitting
                }
                onClick={submitForm}
              >
                Next
              </Button>
            </Grid>
          </Grid>
        </Form>
      )}
    </Formik>
  );
};
