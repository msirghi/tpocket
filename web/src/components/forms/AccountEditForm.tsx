import React, { useState } from 'react';
import { Formik, Form, Field } from 'formik';
import { Button, Grid, MenuItem } from '@material-ui/core';
import { Select, TextField } from 'formik-material-ui';
import { useTranslation } from 'react-i18next';
import ValidationService from '../../utils/ValidationService';
import { useUpdateUserMutation } from '../../generated/graphql';
import { useSnackbar } from 'notistack';
import { AlertMessage } from '../alerts/AlertMessage';
import { AlertType } from '../../commons/enums';
import { useWindowSize } from '../../utils/useWindowSize';

const currencies = [
  {
    value: 'USD',
    label: '$'
  },
  {
    value: 'EUR',
    label: '€'
  },
  {
    value: 'BTC',
    label: '฿'
  },
  {
    value: 'JPY',
    label: '¥'
  }
];

type Props = {
  firstName: string;
  lastName: string;
  monthLimit: number;
  currency: string;
  email: string;
};

type Values = {
  firstName: string;
  lastName: string;
  monthLimit: string;
  currency: string;
  email: string;
};

export const AccountEditForm: React.FC<Props> = ({
  firstName,
  lastName,
  monthLimit,
  currency,
  email
}) => {
  const { t } = useTranslation();
  const [updateUser] = useUpdateUserMutation();
  const [error, setError] = useState(null);
  const { enqueueSnackbar } = useSnackbar();
  const [width] = useWindowSize();

  const onSubmit = async (values: any, setSubmitting: (val: boolean) => void) => {
    const { firstName, lastName, monthLimit, currency } = values;
    try {
      const response = await updateUser({
        variables: {
          currency,
          lastName,
          firstName,
          monthLimit: +monthLimit
        }
      });

      if (response && response.data?.updateUserPreference) {
        enqueueSnackbar(t('messages.accountInfoUpdated'), { variant: 'success' });
      }
      setSubmitting(false);
    } catch (e) {
      setError(e.graphQLErrors.map((x: { message: string }) => x.message)[0]);
      setSubmitting(false);
    }
  };

  return (
    <div>
      <Formik
        initialValues={{
          firstName,
          lastName,
          currency,
          monthLimit,
          email
        }}
        validateOnChange
        validate={(values) => {
          const errors: Partial<Values> = {};
          const { firstName, lastName, monthLimit } = values;

          if (+monthLimit <= 0) {
            errors.monthLimit = t('errors.negativeMonthLimit');
          }

          if (!ValidationService.validateName(firstName)) {
            errors.firstName = t('errors.invalidName');
          }

          if (!ValidationService.validateName(lastName)) {
            errors.lastName = t('errors.invalidName');
          }
          return errors;
        }}
        onSubmit={async (values, { setSubmitting }) => await onSubmit(values, setSubmitting)}
      >
        {({ submitForm, isSubmitting, values, isValid }) => (
          <Form className={'row row-center'}>
            <Grid
              container
              spacing={4}
              xs={width < 768 ? 12 : 6}
              justify='center'
              alignItems='center'
            >
              {error && <AlertMessage type={AlertType.ERROR} message={t('errors.serverError')} />}
              <Grid item xs={12}>
                <Field
                  component={TextField}
                  disabled
                  required
                  data-test='email-field'
                  inputProps={{ 'data-testid': 'firstname-field' }}
                  name='email'
                  label={t('forms.account.email')}
                  fullWidth
                />
              </Grid>

              <Grid item xs={12}>
                <Field
                  component={TextField}
                  required
                  data-test='firstname-field'
                  inputProps={{ 'data-testid': 'firstname-field' }}
                  name='firstName'
                  label={t('forms.account.firstName')}
                  fullWidth
                  autoComplete='given-name'
                />
              </Grid>

              <Grid item xs={12}>
                <Field
                  component={TextField}
                  required
                  data-test='lastname-field'
                  inputProps={{ 'data-testid': 'lastname-field' }}
                  name='lastName'
                  label={t('forms.account.lastName')}
                  fullWidth
                  autoComplete='given-name'
                />
              </Grid>

              <Grid item xs={12}>
                <Field
                  type='number'
                  component={TextField}
                  required
                  data-test='monthlimit-field'
                  inputProps={{ 'data-testid': 'monthlimit-field' }}
                  name='monthLimit'
                  label={t('forms.account.monthLimit')}
                  fullWidth
                />
              </Grid>

              <Grid item xs={12}>
                <Field
                  inputProps={{ 'data-testid': 'currency-selector' }}
                  label={t('forms.additionalInfo.currency')}
                  placeholder={t('forms.account.currency')}
                  fullWidth
                  name='currency'
                  component={Select}
                  defaultValue={values.currency}
                >
                  {currencies.map((option) => (
                    <MenuItem key={option.value} value={option.value}>
                      {option.label}
                    </MenuItem>
                  ))}
                </Field>
              </Grid>

              <Grid item xs={12}>
                <Button
                  data-testid='submit-button'
                  variant='contained'
                  color='primary'
                  disabled={isSubmitting || !isValid}
                  onClick={submitForm}
                  fullWidth
                >
                  {isSubmitting
                    ? t('forms.account.submittingLabel')
                    : t('forms.account.submitButtonLabel')}
                </Button>
              </Grid>
            </Grid>
          </Form>
        )}
      </Formik>
    </div>
  );
};
