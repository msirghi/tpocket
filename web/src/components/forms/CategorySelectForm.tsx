import React from 'react';
import Chip from '@material-ui/core/Chip';
import { startCase } from 'lodash';
import { useTheme } from '@material-ui/core/styles';

const categories = [
  {
    label: 'money'
  },
  {
    label: 'food'
  },
  {
    label: 'technology'
  },
  {
    label: 'presents'
  },
  {
    label: 'subscriptions'
  },
  {
    label: 'transport'
  }
];

type Props = {
  onCategorySelect: (category: string) => void;
  selectedCategories: string[];
};

export const CategorySelectForm: React.FC<Props> = ({ onCategorySelect, selectedCategories }) => {
  const theme = useTheme();

  return (
    <div className='category-select-wrapper'>
      {categories.map((category, idx) => {
        const selected = selectedCategories.includes(category.label);
        return (
          <Chip
            data-testid='chip-option'
            style={{
              backgroundColor: selected ? theme.palette.primary.main : '#f3f3f3',
              color: selected ? '#fff' : 'initial'
            }}
            onClick={() => onCategorySelect(category.label)}
            key={idx}
            label={startCase(category.label)}
          />
        );
      })}
    </div>
  );
};
