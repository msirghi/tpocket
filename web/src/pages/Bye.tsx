import React from 'react'
// import { useByeQuery } from '../generated/graphql';

export const Bye: React.FC<{}> = () => {
  // const { data, loading, error } = useByeQuery({
  //   fetchPolicy: 'network-only'
  // });

  // if (loading) {
  //   return <div>loading...</div>
  // }

  // if (error) {
  //   console.log(error);
  //   return <div>err</div>
  // }

  // if (!data) {
  //   return <div>No data</div>
  // }

  return (<div>
    {/*{ data.bye }*/}
  </div>)

}
