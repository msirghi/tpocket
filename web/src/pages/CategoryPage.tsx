import React, { useEffect, useState } from 'react';
import {
  Category,
  useCategoryByUserQuery,
  useCreateCategoryMutation,
  useDeleteCategoryMutation,
} from '../generated/graphql';
import { CategoryCard } from '../components/category/CategoryCard';
import { Grid } from '@material-ui/core';
import { PageHeader } from '../components/layout/PageHeader';
import {
  CATEGORY_CREATED,
  CATEGORY_DELETE_WARN,
  CATEGORY_DELETED,
  CATEGORY_HEADER_SECONDARY,
} from '../constants';
import AddIcon from '@material-ui/icons/Add';
import SearchIcon from '@material-ui/icons/Search';
import { TwoRowButtons } from '../components/button/TwoRowButtons';
import { CategorySearch } from '../components/category/CategorySearch';
import { AddUpdateCategoryDialog } from '../components/dialog/AddUpdateCategoryDialog';
import { useSnackbar } from 'notistack';
import { ConfirmationDialog } from '../components/dialog/ConfirmationDialog';

type ICategory =
  | Array<{ __typename?: 'Category' } & Pick<Category, 'id' | 'name'>>
  | undefined;

let initialData: ICategory = [];

export const CategoryPage = () => {
  const { enqueueSnackbar } = useSnackbar();
  const { data, loading } = useCategoryByUserQuery();
  const [createCategory] = useCreateCategoryMutation();
  const [deleteCategory] = useDeleteCategoryMutation();

  const [isLoading, setLoading] = useState<boolean>(false);
  const [searchMode, setSearchMode] = useState<boolean>(false);
  const [searchValue, setSearchValue] = useState<string | null>(null);
  const [categoryData, setCategoryData] = useState<ICategory>([]);
  const [creationDialog, setCreationDialog] = useState<boolean>(false);
  const [selectedCategory, setSelectedCategory] = useState<Category | null>(
    null
  );
  const [confirmationDialog, setConfirmationDialog] = useState<boolean>(false);

  useEffect(() => {
    if (searchValue || typeof searchValue === 'string') {
      setCategoryData(
        data?.getCategoryByUser.filter((category) =>
          category.name.includes(searchValue)
        )
      );
    }
  }, [searchValue, data]);

  useEffect(() => {
    setCategoryData(initialData);
    setSearchValue(null);
  }, [searchMode]);

  useEffect(() => {
    initialData = data?.getCategoryByUser;
    setCategoryData(data?.getCategoryByUser);
  }, [data]);

  const initData = (newData: Array<Category> | ICategory) => {
    initialData = newData;
    setCategoryData(newData);
  };

  const createCategoryHandler = async (val: Category) => {
    setLoading(true);
    try {
      const response = await createCategory({ variables: { ...val } });
      if (response && response.data) {
        enqueueSnackbar(CATEGORY_CREATED, { variant: 'success' });
        setCreationDialog(false);
        setLoading(false);
        const newData = [
          ...data?.getCategoryByUser,
          { id: response.data.createCategory.id, name: val.name },
        ];
        initData(newData);
        data?.getCategoryByUser.push({
          id: response.data.createCategory.id,
          name: val.name,
        });
      }
    } catch (e) {
      setLoading(false);
      enqueueSnackbar(e.graphQLErrors[0].message, { variant: 'error' });
    }
  };

  const removeCategory = (category: Category) => {
    data?.getCategoryByUser.splice(
      data?.getCategoryByUser.indexOf(category),
      1
    );
  };

  const deleteCategoryHandler = async (id: number) => {
    setLoading(true);
    try {
      const response = await deleteCategory({ variables: { id } });

      if (response && response.data) {
        enqueueSnackbar(CATEGORY_DELETED, { variant: 'success' });
        setSelectedCategory(null);
        const newData: ICategory = data?.getCategoryByUser.filter(
          (val) => val.id !== id
        );
        initData(newData);
        setConfirmationDialog(false);
        setLoading(false);
        removeCategory(
          data?.getCategoryByUser.find((val) => val.id === id) as Category
        );
      }
    } catch (e) {
      setLoading(false);
      enqueueSnackbar(e.graphQLErrors[0].message, { variant: 'error' });
    }
  };

  if (loading || !categoryData) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <AddUpdateCategoryDialog
        isLoading={isLoading}
        onSubmit={createCategoryHandler}
        open={creationDialog}
        toggleDialog={(val: boolean) => setCreationDialog(val)}
        selectedCategory={selectedCategory}
        deselectCategory={() => setSelectedCategory(null)}
      />

      <ConfirmationDialog
        open={confirmationDialog}
        toggleDialog={setConfirmationDialog}
        onSubmit={() => deleteCategoryHandler(selectedCategory?.id!)}
        isLoading={isLoading}
        title={'Category delete'}
      >
        {CATEGORY_DELETE_WARN}
      </ConfirmationDialog>

      <PageHeader
        primaryText={'Categories'}
        secondaryText={CATEGORY_HEADER_SECONDARY}
      />

      {searchMode ? (
        <CategorySearch
          resultCountShown={searchValue !== null && searchValue.length > 0}
          resultCount={categoryData.length}
          onChange={(e: React.FormEvent<HTMLInputElement>) =>
            setSearchValue(e.currentTarget.value)
          }
          onClose={() => setSearchMode(false)}
        />
      ) : (
        <TwoRowButtons
          buttonWidth={150}
          leftIcon={<SearchIcon />}
          rightIcon={<AddIcon />}
          leftLabel={'Search '}
          rightLabel={'Add'}
          leftClickHandler={() => setSearchMode(true)}
          rightClickHandler={() => setCreationDialog(true)}
        />
      )}

      <Grid container spacing={2} justify={'center'}>
        {categoryData.length ? (
          categoryData.map((value) => (
            <Grid item>
              <CategoryCard
                imageUrl={'https://source.unsplash.com/random/345x140'}
                key={value.id}
                name={value.name}
                onEdit={() => {
                  setSelectedCategory(value);
                  setCreationDialog(true);
                }}
                onRemove={() => {
                  setConfirmationDialog(true);
                  setSelectedCategory(value);
                }}
              />
            </Grid>
          ))
        ) : (
          <div>No category found.</div>
        )}
      </Grid>
    </>
  );
};
