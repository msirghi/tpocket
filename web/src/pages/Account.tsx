import React, { useContext, useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import { AccountEditForm } from '../components/forms/AccountEditForm';
import { PageHeader } from '../components/layout/PageHeader';
import { FullScreenLoader } from '../components/loaders/FullScreenLoader';
import { AccountContext } from '../context/AccountContext';
import { useGetUserInfoQuery } from '../generated/graphql';

export const Account: React.FC = () => {
  const { t } = useTranslation();
  const accountInfo = useContext(AccountContext);

  return (
    <div>
      <PageHeader primaryText={t('account.title')} secondaryText={t('account.description')} />

      <div className='row row-center'>
        <img
          height={100}
          width={100}
          src='https://upload.wikimedia.org/wikipedia/commons/thumb/6/67/Circle-icons-settings.svg/1024px-Circle-icons-settings.svg.png'
        />
      </div>

      <AccountEditForm {...accountInfo.state!} />
    </div>
  );
};
