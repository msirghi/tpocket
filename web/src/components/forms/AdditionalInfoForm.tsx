import React from 'react';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import MenuItem from '@material-ui/core/MenuItem';
import Button from '@material-ui/core/Button';
import { CategorySelectForm } from './CategorySelectForm';
import { Formik, Field, Form } from 'formik';
import { Select, TextField } from 'formik-material-ui';
import { FormHelperText } from '@material-ui/core';
import InputLabel from '@material-ui/core/InputLabel';

type Props = {
  handleNext?: () => void;
};

type Values = {
  currecny: string;
  monthLimit: string;
};

const currencies = [
  {
    value: 'USD',
    label: '$'
  },
  {
    value: 'EUR',
    label: '€'
  },
  {
    value: 'BTC',
    label: '฿'
  },
  {
    value: 'JPY',
    label: '¥'
  }
];

export const AdditionalInfoForm: React.FC<Props> = ({}) => {
  const onSubmit = () => {};

  return (
    <Formik
      initialValues={{
        currency: '',
        monthLimit: ''
      }}
      onSubmit={(values) => {
        console.log('values :>> ', values);
      }}
    >
      {({ submitForm, isSubmitting, values }) => (
        <Form>
          <Typography variant='h6' gutterBottom>
            Additional Info
          </Typography>
          <Grid container spacing={3} className='row row-vert-end'>
            <Grid item xs={12} md={6}>
              <InputLabel shrink htmlFor='currency'>
                Your currency
              </InputLabel>
              <Field
                label='Currency'
                placeholder='Currency'
                fullWidth
                name='currency'
                component={Select}
                defaultValue='none'
              >
                {currencies.map((option) => (
                  <MenuItem key={option.value} value={option.value}>
                    {option.label}
                  </MenuItem>
                ))}
              </Field>
              <FormHelperText>Select your default currency</FormHelperText>
            </Grid>
            <Grid item xs={12} md={6}>
              <Field
                name='monthLimit'
                component={TextField}
                type='number'
                required
                id='limit'
                label='Your month limit'
                fullWidth
                helperText='You will be notified when you reach this limit'
                InputLabelProps={{
                  shrink: true
                }}
              />
            </Grid>

            <Grid item xs={12} md={12}>
              <Typography variant='h6' gutterBottom>
                Expense categories
              </Typography>
            </Grid>
            <Grid item xs={12} md={12} className='disable-padding'>
              <CategorySelectForm />
            </Grid>
            <Grid item xs={12}>
              <Typography variant='subtitle1' gutterBottom>
                * 3 are required
              </Typography>
            </Grid>
          </Grid>
          <Grid item xs={12} className='row row-center'>
            <Button
              variant='contained'
              color='primary'
              disabled={isSubmitting || !values.currency || !values.monthLimit}
              onClick={submitForm}
              fullWidth
            >
              Submit
            </Button>
          </Grid>
        </Form>
      )}
    </Formik>
  );
};
