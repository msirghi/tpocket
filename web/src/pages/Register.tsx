import React, { useEffect, useState } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';
import Paper from '@material-ui/core/Paper';
import Stepper from '@material-ui/core/Stepper';
import Button from '@material-ui/core/Button';
import Step from '@material-ui/core/Step';
import StepLabel from '@material-ui/core/StepLabel';
import Typography from '@material-ui/core/Typography';
import { AboutYouForm } from '../components/forms/AboutYouForm';
import { AdditionalInfoForm } from '../components/forms/AdditionalInfoForm';
import { AES, enc } from 'crypto-js';
import { Link } from 'react-router-dom';
import { useTranslation } from 'react-i18next';

function Copyright() {
  return (
    <Typography variant='body2' color='textSecondary' align='center'>
      {'Copyright Â© '}
      <Link to='https://material-ui.com/'>TPocket</Link> {new Date().getFullYear()}
      {'.'}
    </Typography>
  );
}

const useStyles = makeStyles((theme) => ({
  appBar: {
    position: 'relative'
  },
  layout: {
    width: 'auto',
    marginLeft: theme.spacing(2),
    marginRight: theme.spacing(2),
    [theme.breakpoints.up(600 + theme.spacing(2) * 2)]: {
      width: 600,
      marginLeft: 'auto',
      marginRight: 'auto'
    }
  },
  paper: {
    marginTop: theme.spacing(3),
    marginBottom: theme.spacing(3),
    padding: theme.spacing(2),
    [theme.breakpoints.up(600 + theme.spacing(3) * 2)]: {
      marginTop: theme.spacing(6),
      marginBottom: theme.spacing(6),
      padding: theme.spacing(3)
    }
  },
  stepper: {
    padding: theme.spacing(3, 0, 5)
  },
  buttons: {
    display: 'flex',
    justifyContent: 'flex-end'
  },
  button: {
    marginTop: theme.spacing(3),
    marginLeft: theme.spacing(1)
  }
}));

const steps = ['About You', 'Additional Info', 'Done!'];
let encryptedUserId: string;
let userId;

export const Register = () => {
  const classes = useStyles();
  const [activeStep, setActiveStep] = useState(0);
  const { t } = useTranslation();

  useEffect(() => {
    const tempId = localStorage.getItem('tempId');
    if (tempId) {
      setEncryptedUserId(tempId);
      const decrypted = AES.decrypt(tempId, process.env.REACT_APP_TEMP_ID_KEY!);
      // @ts-ignore
      userId = +decrypted.toString(enc.Utf8);
      setActiveStep(1);
    }
  }, []);

  const setEncryptedUserId = (userId: string) => (encryptedUserId = userId);

  const getStepContent = (step: number) => {
    switch (step) {
      case 0:
        return (
          <AboutYouForm
            handleNext={handleNext}
            setEncryptedUserId={setEncryptedUserId}
            handleRegister={handleRegister}
          />
        );
      case 1:
        return <AdditionalInfoForm handleNext={handleNext} />;
      default:
        throw new Error(t('errors.unknownStep'));
    }
  };

  const handleRegister = (id: number) => (userId = id);

  const handleNext = () => setActiveStep(activeStep + 1);

  return (
    <>
      <CssBaseline />
      <main className={classes.layout}>
        <Paper className={classes.paper}>
          <Typography component='h1' variant='h4' align='center'>
            {t('forms.register.title')}
          </Typography>
          <Stepper activeStep={activeStep} className={classes.stepper}>
            {steps.map((label) => (
              <Step key={label}>
                <StepLabel>{label}</StepLabel>
              </Step>
            ))}
          </Stepper>
          <>
            {activeStep === steps.length - 1 ? (
              <>
                <Typography variant='h5' gutterBottom>
                {t('forms.register.doneTitle')}
                </Typography>
                <div className='row row-center'>
                  <img
                    width='75%'
                    height='75%'
                    src='https://i.pinimg.com/originals/8c/95/7b/8c957b1d5b6262dd171f288eeadcabd7.gif'
                    alt=''
                  ></img>
                </div>
                <Typography variant='subtitle1'>{t('forms.register.doneDescription')}</Typography>
                <Typography variant='subtitle1'>
                  <Button color='primary' fullWidth>
                    <Link to='/login'>Login</Link>
                  </Button>
                </Typography>
              </>
            ) : (
              <>{getStepContent(activeStep)}</>
            )}
          </>
        </Paper>
        <Copyright />
      </main>
    </>
  );
};
