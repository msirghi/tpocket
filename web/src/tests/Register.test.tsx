import { mount, shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloLink } from 'apollo-link';
import { ApolloProvider } from '@apollo/react-hooks';
import toJson from 'enzyme-to-json';
import { BrowserRouter } from 'react-router-dom';
import { Register } from '../pages/Register';
import { AboutYouForm } from '../components/forms/AboutYouForm';
import { AdditionalInfoForm } from '../components/forms/AdditionalInfoForm';
require('jest-localstorage-mock');

describe('Register component', () => {
  let component: ShallowWrapper;
  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: ApolloLink.from([])
  });

  beforeAll(() => {});

  beforeEach(() => {
    component = shallow(<Register />);
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should decrypt on mount', () => {
    jest.mock('crypto-js', () => ({
      AES: {
        decrypt: jest.fn()
      }
    }));
    localStorage.setItem('tempId', 'someTempId');
    mount(
      <ApolloProvider client={client}>
        <BrowserRouter>
          <Register />
        </BrowserRouter>
      </ApolloProvider>
    );
    expect(localStorage.setItem).toHaveBeenLastCalledWith('tempId', 'someTempId');
    expect(localStorage.__STORE__['tempId']).toBe('someTempId');
  });

  it('should return component on valid current step', () => {
    localStorage.removeItem('tempId');
    const wrapper = shallow(
      <ApolloProvider client={client}>
        <BrowserRouter>
          <Register />
        </BrowserRouter>
      </ApolloProvider>
    );
    expect(localStorage.__STORE__['tempId']).toBe(undefined);
    expect(wrapper.find(AboutYouForm)).toHaveLength(0);
    expect(wrapper.find(AdditionalInfoForm)).toHaveLength(0);
  });
});
