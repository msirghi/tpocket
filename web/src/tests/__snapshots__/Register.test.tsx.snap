// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register component should match the snapshot 1`] = `
<Fragment>
  <WithStyles(CssBaseline) />
  <main
    className="makeStyles-layout-2"
  >
    <WithStyles(ForwardRef(Paper))
      className="makeStyles-paper-3"
    >
      <WithStyles(ForwardRef(Typography))
        align="center"
        component="h1"
        variant="h4"
      >
        Registration
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Stepper))
        activeStep={0}
        className="makeStyles-stepper-4"
      >
        <WithStyles(ForwardRef(Step))
          key="About You"
        >
          <WithStyles(ForwardRef(StepLabel))>
            About You
          </WithStyles(ForwardRef(StepLabel))>
        </WithStyles(ForwardRef(Step))>
        <WithStyles(ForwardRef(Step))
          key="Additional Info"
        >
          <WithStyles(ForwardRef(StepLabel))>
            Additional Info
          </WithStyles(ForwardRef(StepLabel))>
        </WithStyles(ForwardRef(Step))>
        <WithStyles(ForwardRef(Step))
          key="Done!"
        >
          <WithStyles(ForwardRef(StepLabel))>
            Done!
          </WithStyles(ForwardRef(StepLabel))>
        </WithStyles(ForwardRef(Step))>
      </WithStyles(ForwardRef(Stepper))>
      <AboutYouForm
        handleNext={[Function]}
        handleRegister={[Function]}
        setEncryptedUserId={[Function]}
      />
    </WithStyles(ForwardRef(Paper))>
    <Copyright />
  </main>
</Fragment>
`;
